src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +114,162 @@ internal class DiagnosticComputer
                 }
             }
 
-            var diagnosticsComputer = new DiagnosticComputer(document, project,
-                solutionChecksum, ideOptions, span, analysisKind, analyzerInfoCache, hostWorkspaceServices);
-            return diagnosticsComputer.GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken);
+            // We perform prioritized execution of diagnostic computation requests based on the
+            // 'highPriority' boolean parameter.
+            //   - High priority requests forces suspension of all the executing normal priority requests,
+            //     which are re-attempted once the high priority request completes.
+            //   - Normal priority requests wait for all the executing high priority requests to complete
+            //     before starting the compute.
+            //   - Suspended normal priority requests are re-attempted in the below loop.
+            bool suspended;
+
+            do
+            {
+                suspended = false;
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Create a linked cancellation source to allow high priority tasks to cancel normal priority tasks.
+                using var cancellationTokenSource = new CancellationTokenSource();
+                using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, cancellationToken);
+
+                // Fetch the cancellation token here to avoid capturing linkedCts in the computeTask lambda as the task may run after linkedCts has been disposed due to cancellation.
+                var linkedCancellationToken = linkedCancellationTokenSource.Token;
+
+                await WaitOrSuspendExecutingTasksAsync(highPriority, cancellationToken).ConfigureAwait(false);
```

Comment:
pull this above all the linked work (Since you're not using hte linked token).  it's def confusing that you did all that, only to not use that token.  I would also recommend extracting logic out to have less overlap of CTs and to make the pieces clearer.

Reply:
Restructured the code with more detailed and step-wise comments.